### PR TITLE
Replace NumberFormatException control flow in NumericVersionFactory.create (#30)

### LINKED ISSUE #30: NumericVersionFactory.create uses NumberFormatException as control flow (hot-path exception storm)
## Summary
`org.octopusden.releng.versions.NumericVersionFactory.create(String)` parses each version segment with `Integer.parseInt(...)` inside a `try { … } catch (NumberFormatException)` and uses the thrown NFE as **control flow** to decide "numeric vs non-numeric" segment. On any version string with non-numeric segments (e.g. suffixed/labelled versions such as `1.2.3-rc1`, `1.2.3.Final`, or `name-1.2.3-456`), this throws — and catches — a `NumberFormatException` per such segment, on every call.

Confirmed from the shipped `versions-api:2.0.10` bytecode:
```
public IVersionInfo create(String);
   138: invokestatic  java/lang/Integer.parseInt:(Ljava/lang/String;)I
  Exception table:
       136  143  146  Class java/lang/NumberFormatException
```

## Impact (measured)
Found while profiling a downstream consumer under a large workload (~82k version resolutions, each calling `create()`): the run produced **~204,840 thrown exceptions**, dominated by `NumberFormatException` (~2 per call) originating here — all thrown and caught **inside `create()`** (they never escape the library).

- Not a throughput bottleneck on its own (the JVM's fast-throw eventually elides stack traces for the hot `parseInt` NFE; the consumer sat at ~4% CPU).
- But it is real wasted allocation/throws on a hot path, and it **pollutes exception monitoring / APM / debugging** for every consumer — hundreds of thousands of "exceptions" that are entirely normal flow.

## Suggested fix
Replace the `parseInt` + `catch NumberFormatException` with a **non-throwing** numeric check — e.g. validate the segment is all digits first (`Character.isDigit` / a manual scan), or a `toIntOrNull`-style parse that returns a sentinel — so the common non-numeric path does not throw. Behaviour (the resulting `IVersionInfo`) should be unchanged; only the exception-as-control-flow is removed.

## Discovery context
Surfaced during JFR exception profiling of a downstream service while investigating an unrelated memory issue. Low-risk efficiency/observability cleanup in the shared `versions-api` library that benefits all consumers (it ships transitively via `releng-lib`).

### PR DESCRIPTION
**Pull request by pgorbachev**
## Summary
Fixes #30.

`NumericVersionFactory.create(String)` parsed each version segment with `Integer.parseInt` inside a `try/catch` and used the thrown `NumberFormatException` as control flow to decide "numeric vs non-numeric". For any labelled version (`1.2.3-rc1`, `1.2.3.Final`, `name-1.2.3-456`) this threw and caught an NFE per non-numeric segment on **every** call — wasted allocation/throws on a hot path and noise in exception monitoring/APM for all downstream consumers of this shared library.

## Change
- Replace `parseInt` + `catch NumberFormatException` with a non-throwing `parseNonNegativeInt` helper that scans the segment for digits and returns `null` for empty, non-numeric, or out-of-range segments.
- Behaviour of the resulting `IVersionInfo` is unchanged — only the exception-as-control-flow is removed.
- Out-of-range all-digit segments (which previously threw NFE on overflow and were skipped) are still skipped, now without throwing.

## Tests
- All 109 existing tests pass (`mvn test`).
- Added `testLabelledSegmentsAreSkipped` and `testOutOfRangeSegmentIsSkipped` to pin the behaviour.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved version parsing robustness by skipping invalid, non-numeric, and labeled segments rather than failing to parse.
  * Added exception-free segment validation, including ignoring oversized values beyond supported integer limits.
  * Preserves existing handling of suffix flags (e.g., `-SNAPSHOT` and `_RC`).
* **Tests**
  * Added coverage for skipping invalid segments, handling out-of-range numeric segments, and edge cases around leading `+` and `int` boundaries.
* **Chores**
  * Updated ignore rules to exclude macOS `.DS_Store` files.

### CODE REVIEW

**Review by fishinitself — DISMISSED** [review]

File: src/main/java/org/octopusden/releng/versions/NumericVersionFactory.java [hunk 1]
```diff
@@ -28,18 +28,39 @@ public IVersionInfo create(String rawVersion) {
...
 
+    /**
+     * Parses a version segment as a non-negative {@code int} without using exceptions for control flow.
+     * Returns {@code null} for empty, non-numeric, or out-of-range segments (the common case for labelled
+     * versions such as {@code 1.2.3-rc1} or {@code 1.2.3.Final}), so the hot path does not throw and catch
+     * a {@link NumberFormatException} per segment on every call.
+     */
+    private static Integer parseNonNegativeInt(String segment) {
+        int length = segment.length();
+        if (length == 0) {
```
Comment (StanislavGro) [inline src/main/java/org/octopusden/releng/versions/NumericVersionFactory.java]: Maybe it's worth adding a check that the length should not exceed 10? If the value is "> Long.MAX_VALUE", then we will get a negative number and the "if (value > Integer.MAX_VALUE)" check will not work and we will end up with an incorrect number at "(int) value"

**Review by StanislavGro — CHANGES_REQUESTED** [review]

File: src/main/java/org/octopusden/releng/versions/NumericVersionFactory.java (hunk 1 above)
Comment (pgorbachev) [reply src/main/java/org/octopusden/releng/versions/NumericVersionFactory.java]: The accumulator cannot reach `Long.MAX_VALUE` here. After every digit we immediately return `null` if `value > Integer.MAX_VALUE`, so before the next iteration `value <= 2_147_483_647`. The next calculation is therefore at most `2_147_483_647 * 10 + 9`, which is still far below `Long.MAX_VALUE`.

The final `(int) value` is reached only if all iterations passed the `Integer.MAX_VALUE` guard, so the cast is safe.

A `length <= 10` check would change existing behavior: `Integer.parseInt` accepts leading zeroes, so values like `00000000002` must still parse as `2`. I added `testIntBoundaryParsing` to cover `Integer.MAX_VALUE`, `2147483648` being skipped, and an 11-character leading-zero value.

**Comment by fishinitself** [comment]
Remove .DS_Store from the commit and add it to .gitignore.

**Review by qu0d — DISMISSED** [review]

**Review by fishinitself — APPROVED** [review]

**Review by StanislavGro — APPROVED** [review]
